File: src/netconsole/models/api/rail_transit_web.py
from __future__ import annotations

from typing import Literal

from pydantic import Field

from netconsole.models.api.common import ApiModel


class RailTransitTaskDTO(ApiModel):
    task_id: str
    status: str = "PENDING"
    action: str
    artifact_id: str = ""
    artifact_name: str = ""
    available: bool = False
    sha256: str = ""
    size_bytes: int = 0
    message: str = ""
    error_message: str = ""
    result_summary: dict[str, object] = Field(default_factory=dict)


class CarNetworkPointRowDTO(ApiModel):
    train_id: str = ""
    train_no: str = ""
    display_name: str = ""
    tc: str = ""
    end: str = ""
    node_name: str = ""
    node_type: str = ""
    device_id: str = ""
    device_name: str = ""
    device_group: str = ""
    station: str = ""
    primary_address: str = ""
    backup_address: str = ""
    ip_vehicle: str = ""
    ip_uplink: str = ""
    ssh_host: str = ""
    vrrp_ip: str = ""
    address_mapping_mode: str = "global"
    primary_address_role: str = ""
    backup_address_role: str = ""
    remark: str = ""


class CarNetworkPointTableDTO(ApiModel):
    rows: list[CarNetworkPointRowDTO] = Field(default_factory=list)
    global_config: dict[str, object] = Field(default_factory=dict)
    locked: bool = False
    revision: str = ""


class CarNetworkPointPreviewRowDTO(ApiModel):
    row_number: int
    status: Literal["valid", "duplicate", "error"]
    key: str = ""
    message: str = ""
    row: CarNetworkPointRowDTO | None = None


class CarNetworkPointPreviewDTO(ApiModel):
    file_name: str
    file_sha256: str
    duplicate_strategy: Literal["replace", "skip", "error"]
    can_apply: bool
    total_count: int
    valid_count: int
    duplicate_count: int
    error_count: int
    rows: list[CarNetworkPointPreviewRowDTO] = Field(default_factory=list)
    result_rows: list[CarNetworkPointRowDTO] = Field(default_factory=list)


class CarNetworkPointTableWriteRequestDTO(ApiModel):
    rows: list[CarNetworkPointRowDTO] = Field(default_factory=list)
    global_config: dict[str, object] = Field(default_factory=dict)
    target_train: dict[str, object] = Field(default_factory=dict)
    overwrite_custom: bool = False
    explicit_confirmation: bool = False
    audit: dict[str, str] = Field(default_factory=dict)
    revision: str = "missing"


class CarNetworkPointTableTransformRequestDTO(ApiModel):
    operation: Literal["apply_mapping", "apply_global", "apply_global_override", "restore_defaults"]
    rows: list[CarNetworkPointRowDTO] = Field(default_factory=list)
    global_config: dict[str, object] = Field(default_factory=dict)


class CarNetworkPointTableExportRequestDTO(ApiModel):
    format: Literal["xlsx", "csv"] = "xlsx"


class OnlineMrReportRequestDTO(ApiModel):
    output_name: str = ""


class OnlineMrParseRequestDTO(ApiModel):
    force_reparse: bool = False


class OnlineMrDeleteRequestDTO(ApiModel):
    expected_session_id: str = Field(min_length=1, max_length=160)
    explicit_confirmation: bool = False


class OnlineMrDesktopLocationDTO(ApiModel):
    target_type: Literal["file", "directory"]
    path: str


class OnlineMrNoteCreateRequestDTO(ApiModel):
    note: str = Field(min_length=1, max_length=500)
    explicit_confirmation: bool = False
    audit: dict[str, str] = Field(default_factory=dict)


class OnlineMrTimelineQueryDTO(ApiModel):
    limit: int = 500
    offset: int = 0


__all__ = [
    "CarNetworkPointPreviewDTO",
    "CarNetworkPointPreviewRowDTO",
    "CarNetworkPointRowDTO",
    "CarNetworkPointTableDTO",
    "CarNetworkPointTableExportRequestDTO",
    "CarNetworkPointTableTransformRequestDTO",
    "CarNetworkPointTableWriteRequestDTO",
    "OnlineMrNoteCreateRequestDTO",
    "OnlineMrParseRequestDTO",
    "OnlineMrDeleteRequestDTO",
    "OnlineMrDesktopLocationDTO",
    "OnlineMrReportRequestDTO",
    "OnlineMrTimelineQueryDTO",
    "RailTransitTaskDTO",
]

File: apps/desktop_renderer/src/types/railTransitWeb.ts
export interface RailTransitTask {
  task_id: string
  status: string
  action: string
  artifact_id: string
  artifact_name: string
  available: boolean
  artifact_state?: string
  artifact_message?: string
  sha256: string
  size_bytes: number
  message: string
  error_message: string
  result_summary: Record<string, unknown>
}

export type OnlineMrParsedDatabaseUpgradeStatus = 'CURRENT' | 'REQUIRED' | 'UPGRADING' | 'FAILED' | 'RAW_DATA_MISSING'

export interface OnlineMrParsedDatabaseEnsureResult {
  status: OnlineMrParsedDatabaseUpgradeStatus
  current_schema_version: number | null
  target_schema_version: number
  missing_capabilities: string[]
  message: string
  retry_suppressed: boolean
  task: RailTransitTask | null
}

export interface OnlineMrMetricPoint {
  timestamp: string | null
  value: number | null
  text_value: string | null
  dimensions: Record<string, unknown>
}

export interface OnlineMrMetricSeries {
  metric_type: string
  series_key: string
  unit: string
  points: OnlineMrMetricPoint[]
  summary: { count: number; minimum: number | null; maximum: number | null; average: number | null }
}

export interface CarNetworkPointRow {
  train_id: string; train_no: string; display_name: string; tc: string; end: string
  node_name: string; node_type: string; device_id: string; device_name: string; device_group: string
  station: string; primary_address: string; backup_address: string; ip_vehicle: string; ip_uplink: string
  ssh_host: string; vrrp_ip: string; address_mapping_mode: string
  primary_address_role: string; backup_address_role: string; remark: string
}

export interface CarNetworkPointTable {
  rows: CarNetworkPointRow[]
  global_config: Record<string, unknown>
  locked: boolean
  revision: string
}

export interface CarNetworkPointPreviewRow {
  row_number: number; status: 'valid' | 'duplicate' | 'error'; key: string; message: string
  row: CarNetworkPointRow | null
}

export interface CarNetworkPointPreview {
  file_name: string; file_sha256: string; duplicate_strategy: 'replace' | 'skip' | 'error'
  can_apply: boolean; total_count: number; valid_count: number; duplicate_count: number; error_count: number
  rows: CarNetworkPointPreviewRow[]; result_rows: CarNetworkPointRow[]
}

export interface OnlineMrTimelineEvent {
  event_id: string; session_id: string; local_time: string | null; device_time: string | null
  source: string; event_type: string; severity: string | null; title: string; payload: Record<string, unknown>
}

export interface MeshImportProfile {
  mr_id: string
}

export interface OnlineTrainRow {
  train_id: string
  train_no: string
  train_name: string
  communication_status: string
  current_mesh_links: number
  active_sessions: number
  warning_count: number
  last_updated_at: string | null
}

export interface OnlineTrainPage {
  items: OnlineTrainRow[]
  total: number
  page: number
  page_size: number
}
